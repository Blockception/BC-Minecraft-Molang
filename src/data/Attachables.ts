import { Data } from "./Data";

/**Molang data for animations controllers*/
export namespace Attachables {
  /**The list of usable variables for this specific type*/
  export const Variables: Data[] = [{id: "is_enchanted"}];
  /**The list of specific contexts usable for this specific type*/
  export const Contexts: Data[] = [{ id: "is_first_person" }, { id: "item_slot" }];
  /**The list of usable temp variables for this specific type*/
  export const Temps: Data[] = [];
}
