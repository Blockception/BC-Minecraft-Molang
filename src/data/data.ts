import { Types } from "bc-minecraft-bedrock-types";

/**The base type of a molang data item*/
export type Data = Types.Identifiable & Types.Documentated;
