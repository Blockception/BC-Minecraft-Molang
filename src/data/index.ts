/*	Auto generated	*/
export * from "./animations";
export * from "./animations-controllers";
export * from "./attachables";
export * from "./blocks";
export * from "./data";
export * from "./entities";
export * from "./features-rules";
export * from "./general";
export * from "./items";
export * from "./molang-data";
export * from "./molang-function";
export * from "./particles";
export * from "./render-controllers";

