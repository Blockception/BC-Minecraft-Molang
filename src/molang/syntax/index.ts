export * from './builder'
export * from './errors'
export * from './nodes'
export * from './operators'
export * from './parse'
export * from './tokens'
export * from './util'