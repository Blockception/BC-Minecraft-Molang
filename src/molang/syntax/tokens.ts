/** Represents a token in the Molang code */
export interface Token {
  type: TokenType;
  value: string;
  position: number;
}

export namespace Token {
  export function oneOfType(t: Pick<Token, "type"> | undefined, ...params: TokenType[]): boolean {
    if (t === undefined) return false;

    return params.includes(t.type);
  }
}

export enum TokenType {
  ArrayAccess,
  Arrow,
  Assignment,
  CloseBrace,
  CloseBracket,
  CloseParen,
  Colon,
  Comma,
  Dot,
  Identifier,
  NamespacedIdentifier,
  NullishCoalescing,
  Number,
  OpenBrace,
  OpenBracket,
  OpenParen,
  Operator,
  QuestionMark,
  Semicolon,
  StringLiteral,
  UnaryOperator,
  EOF,
}

function isDigit(ch: string): boolean {
  const code = ch.charCodeAt(0);
  return code >= 48 && code <= 57; // 0-9
}

function isAlpha(ch: string): boolean {
  const code = ch.charCodeAt(0);
  return (
    (code >= 65 && code <= 90) || // A-Z
    (code >= 97 && code <= 122) || // a-z
    ch === "_"
  );
}

function isAlphaNumeric(ch: string): boolean {
  return isAlpha(ch) || isDigit(ch);
}

/** Tokenizes Molang code into a sequence of tokens */
export function tokenize(input: string): Token[] {
  const tokens: Token[] = [];
  let pos = 0;

  while (pos < input.length) {
    const char = input[pos];

    // Skip whitespace
    switch (char) {
      case " ":
      case "\t":
      case "\r":
      case "\n":
        pos++;
        continue;
    }

    // Numbers
    if (isDigit(char)) {
      let value = "";
      const start = pos;

      while (pos < input.length && (isDigit(input[pos]) || input[pos] === ".")) {
        value += input[pos];
        pos++;
      }

      if (input[pos] === "f") {
        value += input[pos];
        pos++;
      }

      tokens.push({ type: TokenType.Number, value, position: start });
      continue;
    }

    // Identifiers (including namespaced ones like math.sin and minecraft:state)
    if (isAlpha(char)) {
      let value = "";
      const start = pos;

      while (pos < input.length) {
        const ch = input[pos];
        // Allow alphanumeric, dots, underscores, and colons in identifiers
        if (isAlphaNumeric(ch) || ch === "." || ch === ":") {
          value += ch;
          pos++;
        } else {
          break;
        }
      }

      // If it contains a dot, it's a namespaced identifier
      if (value.includes(".")) {
        tokens.push({ type: TokenType.NamespacedIdentifier, value: value.toLowerCase(), position: start });
      } else {
        tokens.push({ type: TokenType.Identifier, value: value.toLowerCase(), position: start });
      }
      continue;
    }

    // Multi-character operators
    const threeCharOp = input.slice(pos, pos + 3);
    if (threeCharOp === "???") {
      tokens.push({ type: TokenType.NullishCoalescing, value: threeCharOp, position: pos });
      pos += 3;
      continue;
    }

    const twoCharOp = input.slice(pos, pos + 2);
    switch (twoCharOp) {
      case "&&":
      case "||":
      case "==":
      case "!=":
      case "<=":
      case ">=":
      case "->":
        tokens.push({ type: TokenType.Operator, value: twoCharOp, position: pos });
        pos += 2;
        continue;
      case "??":
        tokens.push({ type: TokenType.NullishCoalescing, value: twoCharOp, position: pos });
        pos += 2;
        continue;
      default:
    }

    // Single-character operators and punctuation
    switch (char) {
      case "(":
        tokens.push({ type: TokenType.OpenParen, value: char, position: pos });
        break;
      case ")":
        tokens.push({ type: TokenType.CloseParen, value: char, position: pos });
        break;
      case "{":
        tokens.push({ type: TokenType.OpenBrace, value: char, position: pos });
        break;
      case "}":
        tokens.push({ type: TokenType.CloseBrace, value: char, position: pos });
        break;
      case "[":
        tokens.push({ type: TokenType.OpenBracket, value: char, position: pos });
        break;
      case "]":
        tokens.push({ type: TokenType.CloseBracket, value: char, position: pos });
        break;
      case ";":
        tokens.push({ type: TokenType.Semicolon, value: char, position: pos });
        break;
      case "?":
        tokens.push({ type: TokenType.QuestionMark, value: char, position: pos });
        break;
      case ":":
        tokens.push({ type: TokenType.Colon, value: char, position: pos });
        break;
      case ".":
        tokens.push({ type: TokenType.Dot, value: char, position: pos });
        break;
      case "=":
        tokens.push({ type: TokenType.Assignment, value: char, position: pos });
        break;
      case "!":
        tokens.push({ type: TokenType.UnaryOperator, value: char, position: pos });
        break;
      case "+":
      case "-":
      case "*":
      case "/":
      case "<":
      case ">":
        tokens.push({ type: TokenType.Operator, value: char, position: pos });
        break;
      case ",":
        tokens.push({ type: TokenType.Comma, value: char, position: pos });
        break;
      case "'":
      case '"':
        // Handle string literals
        const quote = char;
        const start = pos;
        pos++;
        let value = "";
        while (pos < input.length && input[pos] !== quote) {
          // Capture everything inside quotes including special characters
          value += input[pos];
          pos++;
        }
        if (pos >= input.length) {
          throw new Error(`Unterminated string literal starting at position ${start}`);
        }
        pos++; // Skip closing quote
        tokens.push({
          type: TokenType.StringLiteral,
          value: value, // Use accumulated value instead of slicing
          position: start,
        });
        continue; // Use continue instead of break to avoid the pos++ at the end of switch
      default:
        throw new Error(`Unexpected character at position ${pos}: ${char}`);
    }
    pos++;
    continue;
  }

  // Always add an EOF token at the end
  tokens.push({ type: TokenType.EOF, value: " ", position: pos });

  // Check for unary operators
  tokens.forEach(checkUnaryOperators);

  return tokens;
}

function checkUnaryOperators(item: Token, index: number, items: Token[]) {
  if (item.type === TokenType.Identifier && item.value === "return") {
    item.type = TokenType.UnaryOperator;
    return;
  }
  if (item.type !== TokenType.Operator && item.value !== "-") return;

  if (index === 0) {
    item.type = TokenType.UnaryOperator;
    return;
  }
  const previous = items[index - 1];
  switch (previous.type) {
    case TokenType.Comma:
    case TokenType.Operator:
    case TokenType.OpenBrace:
    case TokenType.OpenBracket:
    case TokenType.OpenParen:
      item.type = TokenType.UnaryOperator;
      break;
    default:
  }
}
