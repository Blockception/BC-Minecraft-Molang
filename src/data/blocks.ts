import { Data } from "./data";

/**Molang data for blocks*/
export namespace Blocks {
  /**The list of usable variables for this specific type*/
  export const Variables: Data[] = [];
  /**The list of specific contexts usable for this specific type*/
  export const Contexts: Data[] = [];
    /**The list of usable temp variables for this specific type*/
  export const Temps: Data[] = [];
}
