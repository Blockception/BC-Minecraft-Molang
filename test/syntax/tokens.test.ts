import { tokenize, TokenType } from "../../src/molang/syntax/tokens";
import { valid_syntaxes } from "../data/dataset";

describe("molang - syntax", () => {
  describe("should be able to parse and match the syntax tree generated", () => {
    test.each(valid_syntaxes)("%#. %s", (s) => {
      const n = tokenize(s);
      expect(n.map(item => `${item.value} ${TokenType[item.type]}`)).toMatchSnapshot();
    });
  });
});
