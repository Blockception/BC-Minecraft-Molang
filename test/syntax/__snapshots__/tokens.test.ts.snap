// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`molang - syntax should be able to parse and match the syntax tree generated 0. variable.agent.anim_state == 2 1`] = `
[
  "variable.agent.anim_state NamespacedIdentifier",
  "== Operator",
  "2 Number",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 1. variable.agent.anim_state > 2 1`] = `
[
  "variable.agent.anim_state NamespacedIdentifier",
  "> Operator",
  "2 Number",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 2. variable.agent.anim_state < 2 1`] = `
[
  "variable.agent.anim_state NamespacedIdentifier",
  "< Operator",
  "2 Number",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 3. variable.agent.anim_state != 2 1`] = `
[
  "variable.agent.anim_state NamespacedIdentifier",
  "!= Operator",
  "2 Number",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 4. variable.agent.anim_state >= 2 1`] = `
[
  "variable.agent.anim_state NamespacedIdentifier",
  ">= Operator",
  "2 Number",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 5. variable.agent.anim_state <= 2 1`] = `
[
  "variable.agent.anim_state NamespacedIdentifier",
  "<= Operator",
  "2 Number",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 6. !variable.swaying && query.all_animations_finished 1`] = `
[
  "! UnaryOperator",
  "variable.swaying NamespacedIdentifier",
  "&& Operator",
  "query.all_animations_finished NamespacedIdentifier",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 7. v.agent.anim_state == 2 1`] = `
[
  "v.agent.anim_state NamespacedIdentifier",
  "== Operator",
  "2 Number",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 8. v.agent.anim_state > 2 1`] = `
[
  "v.agent.anim_state NamespacedIdentifier",
  "> Operator",
  "2 Number",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 9. v.agent.anim_state < 2 1`] = `
[
  "v.agent.anim_state NamespacedIdentifier",
  "< Operator",
  "2 Number",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 10. v.agent.anim_state != 2 1`] = `
[
  "v.agent.anim_state NamespacedIdentifier",
  "!= Operator",
  "2 Number",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 11. v.agent.anim_state >= 2 1`] = `
[
  "v.agent.anim_state NamespacedIdentifier",
  ">= Operator",
  "2 Number",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 12. v.agent.anim_state <= 2 1`] = `
[
  "v.agent.anim_state NamespacedIdentifier",
  "<= Operator",
  "2 Number",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 13. !v.swaying && q.all_animations_finished 1`] = `
[
  "! UnaryOperator",
  "v.swaying NamespacedIdentifier",
  "&& Operator",
  "q.all_animations_finished NamespacedIdentifier",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 14. 1.5 - math.pow(1.5, 1.0 - 015.0 * query.modified_move_speed) * query.camera_rotation(0) 1`] = `
[
  "1.5 Number",
  "- Operator",
  "math.pow NamespacedIdentifier",
  "( OpenParen",
  "1.5 Number",
  ", Comma",
  "1.0 Number",
  "- Operator",
  "015.0 Number",
  "* Operator",
  "query.modified_move_speed NamespacedIdentifier",
  ") CloseParen",
  "* Operator",
  "query.camera_rotation NamespacedIdentifier",
  "( OpenParen",
  "0 Number",
  ") CloseParen",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 15. math.clamp(math.sqrt(math.pow(query.position_delta(0), 2.0) + math.pow(query.position_delta(2), 2.0)), 0.0, 0.1) 1`] = `
[
  "math.clamp NamespacedIdentifier",
  "( OpenParen",
  "math.sqrt NamespacedIdentifier",
  "( OpenParen",
  "math.pow NamespacedIdentifier",
  "( OpenParen",
  "query.position_delta NamespacedIdentifier",
  "( OpenParen",
  "0 Number",
  ") CloseParen",
  ", Comma",
  "2.0 Number",
  ") CloseParen",
  "+ Operator",
  "math.pow NamespacedIdentifier",
  "( OpenParen",
  "query.position_delta NamespacedIdentifier",
  "( OpenParen",
  "2 Number",
  ") CloseParen",
  ", Comma",
  "2.0 Number",
  ") CloseParen",
  ") CloseParen",
  ", Comma",
  "0.0 Number",
  ", Comma",
  "0.1 Number",
  ") CloseParen",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 16. -175 * math.sin(variable.attack_time * 180) 1`] = `
[
  "- UnaryOperator",
  "175 Number",
  "* Operator",
  "math.sin NamespacedIdentifier",
  "( OpenParen",
  "variable.attack_time NamespacedIdentifier",
  "* Operator",
  "180 Number",
  ") CloseParen",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 17. math.sin(variable.attack_time * 180) * -175 1`] = `
[
  "math.sin NamespacedIdentifier",
  "( OpenParen",
  "variable.attack_time NamespacedIdentifier",
  "* Operator",
  "180 Number",
  ") CloseParen",
  "* Operator",
  "- UnaryOperator",
  "175 Number",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 18. -5.0 * Math.cos(297.9380535 * query.anim_time) - 5.0 1`] = `
[
  "- UnaryOperator",
  "5.0 Number",
  "* Operator",
  "math.cos NamespacedIdentifier",
  "( OpenParen",
  "297.9380535 Number",
  "* Operator",
  "query.anim_time NamespacedIdentifier",
  ") CloseParen",
  "- Operator",
  "5.0 Number",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 19. variable.is_enchanted ? Material.enchanted : Material.default 1`] = `
[
  "variable.is_enchanted NamespacedIdentifier",
  "? QuestionMark",
  "material.enchanted NamespacedIdentifier",
  ": Colon",
  "material.default NamespacedIdentifier",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 20. variable.has_trim ? variable.trim_path : Texture.default 1`] = `
[
  "variable.has_trim NamespacedIdentifier",
  "? QuestionMark",
  "variable.trim_path NamespacedIdentifier",
  ": Colon",
  "texture.default NamespacedIdentifier",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 21. !variable.has_trim 1`] = `
[
  "! UnaryOperator",
  "variable.has_trim NamespacedIdentifier",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 22. array.skins[query.property('minecraft:has_nectar') + query.is_angry * 2] 1`] = `
[
  "array.skins NamespacedIdentifier",
  "[ OpenBracket",
  "query.property NamespacedIdentifier",
  "( OpenParen",
  "minecraft:has_nectar StringLiteral",
  ") CloseParen",
  "+ Operator",
  "query.is_angry NamespacedIdentifier",
  "* Operator",
  "2 Number",
  "] CloseBracket",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 23. variable.x_head_rotation = query.target_x_rotation * 32.5; 1`] = `
[
  "variable.x_head_rotation NamespacedIdentifier",
  "= Assignment",
  "query.target_x_rotation NamespacedIdentifier",
  "* Operator",
  "32.5 Number",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 24. variable.x_head_rotation = math.clamp(query.target_x_rotation, -32.5, 32.5); 1`] = `
[
  "variable.x_head_rotation NamespacedIdentifier",
  "= Assignment",
  "math.clamp NamespacedIdentifier",
  "( OpenParen",
  "query.target_x_rotation NamespacedIdentifier",
  ", Comma",
  "- UnaryOperator",
  "32.5 Number",
  ", Comma",
  "32.5 Number",
  ") CloseParen",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 25. variable.y_head_rotation = math.clamp(query.target_y_rotation, -22.5, 25); 1`] = `
[
  "variable.y_head_rotation NamespacedIdentifier",
  "= Assignment",
  "math.clamp NamespacedIdentifier",
  "( OpenParen",
  "query.target_y_rotation NamespacedIdentifier",
  ", Comma",
  "- UnaryOperator",
  "22.5 Number",
  ", Comma",
  "25 Number",
  ") CloseParen",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 26. variable.is_rolled_up = query.property('minecraft:armadillo_state') != 'unrolled'; 1`] = `
[
  "variable.is_rolled_up NamespacedIdentifier",
  "= Assignment",
  "query.property NamespacedIdentifier",
  "( OpenParen",
  "minecraft:armadillo_state StringLiteral",
  ") CloseParen",
  "!= Operator",
  "unrolled StringLiteral",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 27. variable.is_rolled_up = (variable.unrolling_time == 0.0 || variable.unrolling_time <= 1.25 ); 1`] = `
[
  "variable.is_rolled_up NamespacedIdentifier",
  "= Assignment",
  "( OpenParen",
  "variable.unrolling_time NamespacedIdentifier",
  "== Operator",
  "0.0 Number",
  "|| Operator",
  "variable.unrolling_time NamespacedIdentifier",
  "<= Operator",
  "1.25 Number",
  ") CloseParen",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 28. variable.rolled_up_time = variable.is_rolled_up ? ((variable.rolled_up_time ?? 0.0) + query.delta_time) : 0.0; 1`] = `
[
  "variable.rolled_up_time NamespacedIdentifier",
  "= Assignment",
  "variable.is_rolled_up NamespacedIdentifier",
  "? QuestionMark",
  "( OpenParen",
  "( OpenParen",
  "variable.rolled_up_time NamespacedIdentifier",
  "?? NullishCoalescing",
  "0.0 Number",
  ") CloseParen",
  "+ Operator",
  "query.delta_time NamespacedIdentifier",
  ") CloseParen",
  ": Colon",
  "0.0 Number",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 29. variable.is_peeking = query.property('minecraft:armadillo_state') == 'rolled_up_peeking'; 1`] = `
[
  "variable.is_peeking NamespacedIdentifier",
  "= Assignment",
  "query.property NamespacedIdentifier",
  "( OpenParen",
  "minecraft:armadillo_state StringLiteral",
  ") CloseParen",
  "== Operator",
  "rolled_up_peeking StringLiteral",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 30. variable.is_unrolling = query.property('minecraft:armadillo_state') == 'rolled_up_unrolling'; 1`] = `
[
  "variable.is_unrolling NamespacedIdentifier",
  "= Assignment",
  "query.property NamespacedIdentifier",
  "( OpenParen",
  "minecraft:armadillo_state StringLiteral",
  ") CloseParen",
  "== Operator",
  "rolled_up_unrolling StringLiteral",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 31. variable.unrolling_time = variable.is_unrolling ? ((variable.unrolling_time ?? 0.0) + query.delta_time) : 0.0; 1`] = `
[
  "variable.unrolling_time NamespacedIdentifier",
  "= Assignment",
  "variable.is_unrolling NamespacedIdentifier",
  "? QuestionMark",
  "( OpenParen",
  "( OpenParen",
  "variable.unrolling_time NamespacedIdentifier",
  "?? NullishCoalescing",
  "0.0 Number",
  ") CloseParen",
  "+ Operator",
  "query.delta_time NamespacedIdentifier",
  ") CloseParen",
  ": Colon",
  "0.0 Number",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 32. variable.use_rolled_up_model = variable.rolled_up_time >= 0.2083 && (variable.unrolling_time == 0.0 || variable.unrolling_time <= 1.25 ); 1`] = `
[
  "variable.use_rolled_up_model NamespacedIdentifier",
  "= Assignment",
  "variable.rolled_up_time NamespacedIdentifier",
  ">= Operator",
  "0.2083 Number",
  "&& Operator",
  "( OpenParen",
  "variable.unrolling_time NamespacedIdentifier",
  "== Operator",
  "0.0 Number",
  "|| Operator",
  "variable.unrolling_time NamespacedIdentifier",
  "<= Operator",
  "1.25 Number",
  ") CloseParen",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 33. variable.walking = query.modified_move_speed > 0.01 && !variable.is_rolled_up; 1`] = `
[
  "variable.walking NamespacedIdentifier",
  "= Assignment",
  "query.modified_move_speed NamespacedIdentifier",
  "> Operator",
  "0.01 Number",
  "&& Operator",
  "! UnaryOperator",
  "variable.is_rolled_up NamespacedIdentifier",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 34. variable.walk_anim_time_update = query.anim_time + math.min(3.0, math.lerp(2.0, 5.0, query.modified_move_speed)) * query.delta_time; 1`] = `
[
  "variable.walk_anim_time_update NamespacedIdentifier",
  "= Assignment",
  "query.anim_time NamespacedIdentifier",
  "+ Operator",
  "math.min NamespacedIdentifier",
  "( OpenParen",
  "3.0 Number",
  ", Comma",
  "math.lerp NamespacedIdentifier",
  "( OpenParen",
  "2.0 Number",
  ", Comma",
  "5.0 Number",
  ", Comma",
  "query.modified_move_speed NamespacedIdentifier",
  ") CloseParen",
  ") CloseParen",
  "* Operator",
  "query.delta_time NamespacedIdentifier",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 35. variable.walk_anim_time_update = query.anim_time + math.min(3.0, math.lerp(2.0, 5.0, query.modified_move_speed)) * query.delta_time;variable.y_head_rotation = math.clamp(query.target_y_rotation, -22.5, 25); 1`] = `
[
  "variable.walk_anim_time_update NamespacedIdentifier",
  "= Assignment",
  "query.anim_time NamespacedIdentifier",
  "+ Operator",
  "math.min NamespacedIdentifier",
  "( OpenParen",
  "3.0 Number",
  ", Comma",
  "math.lerp NamespacedIdentifier",
  "( OpenParen",
  "2.0 Number",
  ", Comma",
  "5.0 Number",
  ", Comma",
  "query.modified_move_speed NamespacedIdentifier",
  ") CloseParen",
  ") CloseParen",
  "* Operator",
  "query.delta_time NamespacedIdentifier",
  "; Semicolon",
  "variable.y_head_rotation NamespacedIdentifier",
  "= Assignment",
  "math.clamp NamespacedIdentifier",
  "( OpenParen",
  "query.target_y_rotation NamespacedIdentifier",
  ", Comma",
  "- UnaryOperator",
  "22.5 Number",
  ", Comma",
  "25 Number",
  ") CloseParen",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 36. variable.x_head_rotation = math.clamp(query.target_x_rotation, -32.5, 32.5); 1`] = `
[
  "variable.x_head_rotation NamespacedIdentifier",
  "= Assignment",
  "math.clamp NamespacedIdentifier",
  "( OpenParen",
  "query.target_x_rotation NamespacedIdentifier",
  ", Comma",
  "- UnaryOperator",
  "32.5 Number",
  ", Comma",
  "32.5 Number",
  ") CloseParen",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 37. variable.y_head_rotation = math.clamp(query.target_y_rotation, -22.5, 25); 1`] = `
[
  "variable.y_head_rotation NamespacedIdentifier",
  "= Assignment",
  "math.clamp NamespacedIdentifier",
  "( OpenParen",
  "query.target_y_rotation NamespacedIdentifier",
  ", Comma",
  "- UnaryOperator",
  "22.5 Number",
  ", Comma",
  "25 Number",
  ") CloseParen",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 38. variable.is_rolled_up = q.property('minecraft:armadillo_state') != 'unrolled'; 1`] = `
[
  "variable.is_rolled_up NamespacedIdentifier",
  "= Assignment",
  "q.property NamespacedIdentifier",
  "( OpenParen",
  "minecraft:armadillo_state StringLiteral",
  ") CloseParen",
  "!= Operator",
  "unrolled StringLiteral",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 39. variable.rolled_up_time = variable.is_rolled_up ? ((variable.rolled_up_time ?? 0.0) + q.delta_time) : 0.0; 1`] = `
[
  "variable.rolled_up_time NamespacedIdentifier",
  "= Assignment",
  "variable.is_rolled_up NamespacedIdentifier",
  "? QuestionMark",
  "( OpenParen",
  "( OpenParen",
  "variable.rolled_up_time NamespacedIdentifier",
  "?? NullishCoalescing",
  "0.0 Number",
  ") CloseParen",
  "+ Operator",
  "q.delta_time NamespacedIdentifier",
  ") CloseParen",
  ": Colon",
  "0.0 Number",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 40. variable.is_peeking = q.property('minecraft:armadillo_state') == 'rolled_up_peeking'; 1`] = `
[
  "variable.is_peeking NamespacedIdentifier",
  "= Assignment",
  "q.property NamespacedIdentifier",
  "( OpenParen",
  "minecraft:armadillo_state StringLiteral",
  ") CloseParen",
  "== Operator",
  "rolled_up_peeking StringLiteral",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 41. variable.is_unrolling = q.property('minecraft:armadillo_state') == 'rolled_up_unrolling'; 1`] = `
[
  "variable.is_unrolling NamespacedIdentifier",
  "= Assignment",
  "q.property NamespacedIdentifier",
  "( OpenParen",
  "minecraft:armadillo_state StringLiteral",
  ") CloseParen",
  "== Operator",
  "rolled_up_unrolling StringLiteral",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 42. variable.unrolling_time = variable.is_unrolling ? ((variable.unrolling_time ?? 0.0) + q.delta_time) : 0.0; 1`] = `
[
  "variable.unrolling_time NamespacedIdentifier",
  "= Assignment",
  "variable.is_unrolling NamespacedIdentifier",
  "? QuestionMark",
  "( OpenParen",
  "( OpenParen",
  "variable.unrolling_time NamespacedIdentifier",
  "?? NullishCoalescing",
  "0.0 Number",
  ") CloseParen",
  "+ Operator",
  "q.delta_time NamespacedIdentifier",
  ") CloseParen",
  ": Colon",
  "0.0 Number",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 43. variable.walking = q.modified_move_speed > 0.01 && !variable.is_rolled_up; 1`] = `
[
  "variable.walking NamespacedIdentifier",
  "= Assignment",
  "q.modified_move_speed NamespacedIdentifier",
  "> Operator",
  "0.01 Number",
  "&& Operator",
  "! UnaryOperator",
  "variable.is_rolled_up NamespacedIdentifier",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 44. variable.walk_anim_time_update = q.anim_time + math.min(3.0, math.lerp(2.0, 5.0, q.modified_move_speed)) * q.delta_time; 1`] = `
[
  "variable.walk_anim_time_update NamespacedIdentifier",
  "= Assignment",
  "q.anim_time NamespacedIdentifier",
  "+ Operator",
  "math.min NamespacedIdentifier",
  "( OpenParen",
  "3.0 Number",
  ", Comma",
  "math.lerp NamespacedIdentifier",
  "( OpenParen",
  "2.0 Number",
  ", Comma",
  "5.0 Number",
  ", Comma",
  "q.modified_move_speed NamespacedIdentifier",
  ") CloseParen",
  ") CloseParen",
  "* Operator",
  "q.delta_time NamespacedIdentifier",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 45. variable.walk_anim_time_update = q.anim_time + math.min(3.0, math.lerp(2.0, 5.0, q.modified_move_speed)) * q.delta_time;variable.y_head_rotation = math.clamp(q.target_y_rotation, -22.5, 25); 1`] = `
[
  "variable.walk_anim_time_update NamespacedIdentifier",
  "= Assignment",
  "q.anim_time NamespacedIdentifier",
  "+ Operator",
  "math.min NamespacedIdentifier",
  "( OpenParen",
  "3.0 Number",
  ", Comma",
  "math.lerp NamespacedIdentifier",
  "( OpenParen",
  "2.0 Number",
  ", Comma",
  "5.0 Number",
  ", Comma",
  "q.modified_move_speed NamespacedIdentifier",
  ") CloseParen",
  ") CloseParen",
  "* Operator",
  "q.delta_time NamespacedIdentifier",
  "; Semicolon",
  "variable.y_head_rotation NamespacedIdentifier",
  "= Assignment",
  "math.clamp NamespacedIdentifier",
  "( OpenParen",
  "q.target_y_rotation NamespacedIdentifier",
  ", Comma",
  "- UnaryOperator",
  "22.5 Number",
  ", Comma",
  "25 Number",
  ") CloseParen",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 46. variable.state = variable.is_rolled_up ? 1 : 0; 1`] = `
[
  "variable.state NamespacedIdentifier",
  "= Assignment",
  "variable.is_rolled_up NamespacedIdentifier",
  "? QuestionMark",
  "1 Number",
  ": Colon",
  "0 Number",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 47. variable.state = variable.is_rolled_up ? 2 * (5 + variable.state) : 3 * 1 + variable.foo; 1`] = `
[
  "variable.state NamespacedIdentifier",
  "= Assignment",
  "variable.is_rolled_up NamespacedIdentifier",
  "? QuestionMark",
  "2 Number",
  "* Operator",
  "( OpenParen",
  "5 Number",
  "+ Operator",
  "variable.state NamespacedIdentifier",
  ") CloseParen",
  ": Colon",
  "3 Number",
  "* Operator",
  "1 Number",
  "+ Operator",
  "variable.foo NamespacedIdentifier",
  "; Semicolon",
  "  EOF",
]
`;

exports[`molang - syntax should be able to parse and match the syntax tree generated 48. variable.state = variable.is_rolled_up ? variable.rolled_up_time >= 0.2083 ? 1 : 2 : 3; 1`] = `
[
  "variable.state NamespacedIdentifier",
  "= Assignment",
  "variable.is_rolled_up NamespacedIdentifier",
  "? QuestionMark",
  "variable.rolled_up_time NamespacedIdentifier",
  ">= Operator",
  "0.2083 Number",
  "? QuestionMark",
  "1 Number",
  ": Colon",
  "2 Number",
  ": Colon",
  "3 Number",
  "; Semicolon",
  "  EOF",
]
`;
